topic/timeseries/timeseries-anomaly-detection.ipynb

```diff
@@ -0,0 +1,732 @@
+    "\n",
+    "Since algorithms cannot directly consume date or timestamp data, we will extract the features from the timestamp and will drop the actual timestamp column before training models."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<div>\n",
+       "<style scoped>\n",
+       "    .dataframe tbody tr th:only-of-type {\n",
+       "        vertical-align: middle;\n",
+       "    }\n",
+       "\n",
+       "    .dataframe tbody tr th {\n",
+       "        vertical-align: top;\n",
+       "    }\n",
+       "\n",
+       "    .dataframe thead th {\n",
+       "        text-align: right;\n",
+       "    }\n",
+       "</style>\n",
+       "<table border=\"1\" class=\"dataframe\">\n",
+       "  <thead>\n",
+       "    <tr style=\"text-align: right;\">\n",
+       "      <th></th>\n",
+       "      <th>avg_value</th>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>timestamp</th>\n",
+       "      <th></th>\n",
+       "    </tr>\n",
+       "  </thead>\n",
+       "  <tbody>\n",
+       "    <tr>\n",
+       "      <th>2013-12-02 21:15:00</th>\n",
+       "      <td>73.967322</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>2013-12-02 21:20:00</th>\n",
+       "      <td>74.935882</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>2013-12-02 21:25:00</th>\n",
+       "      <td>76.124162</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>2013-12-02 21:30:00</th>\n",
+       "      <td>78.140707</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>2013-12-02 21:35:00</th>\n",
+       "      <td>79.329836</td>\n",
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "                     avg_value\n",
+       "timestamp                     \n",
+       "2013-12-02 21:15:00  73.967322\n",
+       "2013-12-02 21:20:00  74.935882\n",
+       "2013-12-02 21:25:00  76.124162\n",
+       "2013-12-02 21:30:00  78.140707\n",
+       "2013-12-02 21:35:00  79.329836"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "query = \"SELECT DATE_BIN('5 min'::INTERVAL, timestamp, 0) AS timestamp, AVG(value) AS avg_value FROM machine_data GROUP BY timestamp ORDER BY timestamp ASC;\"  \n",
+    "\n",
+    "with engine.connect() as conn:\n",
+    "    df = pd.read_sql(sql=sa.text(query), con=conn)\n",
+    "\n",
+    "df['timestamp'] = df['timestamp'].transform(lambda x: datetime.fromtimestamp(x/1000))\n",
+    "\n",
+    "df = df.set_index('timestamp')\n",
```

Comment:
Hi Karyn. Thank you for adjusting this spot to use pandas directly for reading the data. Now, we only have another question out of curiosity, possibly relating to that note:

- https://github.com/crate/crate-python/discussions/589

Is it a common pattern to apply that transformation on the timestamp column, and to define it to be an index on the data frame? For us, it looks like this is the case, that's why we are constantly thinking about improving UX on this matter, even if it would be about getting rid of just two lines of code, or four, depending on how it would be shaped.

So, what do you think about having such an interface for future variants of the same thing?
```python
cratedb = CrateDBAdapter(engine=engine)
df = cratedb.query_df(sql=query, timestamp_columns=["timestamp"])
```

Reply:
Sure, why not? As long ad it is not mandatory to do this transformation, since this is specific to time series.